### PR TITLE
Initial primal slice operator

### LINKED ISSUE #1579: Add a `slice` operator to primal
PR https://github.com/LLNL/axom/pull/1560 has a locally defined operator that slices a `Tetrahedron` with a `Plane`, resulting in a (possibly empty) polygon with 0, 3, or 4 vertices.

We should move this `primal`

See: https://github.com/LLNL/axom/pull/1560#discussion_r2089973499

### PR DESCRIPTION
**Pull request by BradWhitlock**
This PR resolves #1579 by relocating code from a quest test into primal as a slice operator.

* Added segment-segment intersection
* Added tests
* Updated RELEASE-NOTES.md

### CODE REVIEW

File: src/axom/primal/operators/intersect.hpp [hunk 1]
```diff
@@ -44,6 +44,23 @@ namespace axom
 {
 namespace primal
 {
+
+/*!
+ * \brief Determines if two 3D segments intersect.
+ * \param [in] P A 3D line segment
+ * \param [in] Q A 3D line segment
+ * \param [out] intersection Intersection point of P and Q.
+ * \return true iff P intersects with Q, otherwise, false.
```
Comment (kennyweiss) [inline src/axom/primal/operators/intersect.hpp:57]: Please add a note about what happens if P and Q overlap in a line segment (i.e. they are parallel and overlapping).

Please also mention that the `intersection` point should be ignored when `intersect` returns `false`

File: src/axom/primal/tests/primal_intersect.cpp [hunk 2]
```diff
@@ -281,6 +281,42 @@ TEST(primal_intersect, more_ray_segment_intersection)
   }
 }
 
+TEST(primal_intersect, segment_segment_intersection)
+{
+  using Point3D = primal::Point<double, 3>;
+  using Segment3D = primal::Segment<double, 3>;
+
+  Point3D intersection;
+
```
Comment (kennyweiss) [inline src/axom/primal/tests/primal_intersect.cpp:291]: Please add a test for the two segments intersecting at:
* an endpoint of segment 1 that coincides with an endpoint of segment 2
* an endpoint of segment 1 that coincides with an internal point of segment 2

And also tests for overlapping segments (once you've decided how to handle this case):
* both segments are the same
* both segments are the same, but the second swaps endpoints
* partial overlap b/w the segments -- e.g.  if points A, B, C and D are collinear, segment 1 is (A,C) and segment 2 is (B,D)

File: src/axom/primal/tests/primal_slice.cpp
Comment (kennyweiss) [inline src/axom/primal/tests/primal_slice.cpp:1]: Please add tests on the "edge" cases, where the plane intersects the tet on:
* a vertex of the tet
* an edge of the tet
* a face of the tet

(apologies if they're already there -- doesn't look like it, but I could be missing something

File: src/axom/primal/operators/slice.hpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+/*!
+ * \brief Slices a 3D tetrahedron with a plane and returns the resulting
+ *        polygon.
+ *
+ * \param [in] tet The tetrahedron to slice
+ * \param [in] plane The slicing plane
+ *
+ * \return The polygon obtained from slicing a tetrahedron with a plane.
+ */
```
Comment (kennyweiss) [inline src/axom/primal/operators/slice.hpp:41]: Please discuss what the result should be for edge cases -- where the plane-tet intersection is along a vertex of edge of the tet.  (i.e. a degenerate polygon).

Please also mention the orientation of the resulting polygon -- presumably, it has the same normal as the plane.

File: src/axom/primal/operators/detail/intersect_impl.hpp [hunk 4]
```diff
@@ -134,6 +134,171 @@ bool TriangleIntersection2D(const Triangle2& t1, const Triangle2& t2, bool inclu
 
 //------------------------------ IMPLEMENTATIONS ------------------------------
 
+/*!
+ * \brief Intersect 2 3D line segments and return true if they intersect.
+ *
+ * \param P The first line segment.
+ * \param Q The second line segment.
+ * \param[out] intersection The intersection point where the segments intersect.
```
Comment (kennyweiss) [inline src/axom/primal/operators/detail/intersect_impl.hpp:142]: Please mention how `EPS` is used

File: src/axom/primal/operators/detail/slice_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    Segment<T, 3> seg1(intersectionPolygon[0], intersectionPolygon[1]);
+    Segment<T, 3> seg2(intersectionPolygon[2], intersectionPolygon[3]);
+    Point<T, 3> sp;
+    if(!primal::intersect(seg1, seg2, sp))
+    {
+      axom::utilities::swap(intersectionPolygon[2], intersectionPolygon[3]);
+    }
+  }
+  return intersectionPolygon;
+}
```
Comment (kennyweiss) [inline src/axom/primal/operators/detail/slice_impl.hpp:132]: Should we also ensure that the orientation of the Polygon matches that of the Plane?

**Review by kennyweiss — CHANGES_REQUESTED** [review]
Thanks @BradWhitlock -- I asked for a bunch of edge-case tests and checks.
Probably the most important is that the orientation of the polygon should match the plane, since, presumably, that would be needed for applications.

File: src/axom/primal/operators/intersect.hpp (hunk 1 above)
Comment (BradWhitlock) [reply src/axom/primal/operators/intersect.hpp:57]: The current interface isn't good for P and Q segments overlapping in that way. Would it make more sense to have the intersection output be another Segment that could represent that and in the general cases where P and Q intersect at a point, just make the output segment points the same?

Comment (kennyweiss) [reply src/axom/primal/operators/intersect.hpp:57]: Seems like this case does not come up in the plane-tet slicing.

I guess it depends on what we're using it for. Could be that `intersection` is an arbitrary point in the intersection when they overlap (?).

I think it's probably fine as long as you document the expected behavior.

Comment (BradWhitlock) [reply src/axom/primal/operators/intersect.hpp:57]: I kept the interface the same and documented what happens if P,Q overlap as a line segment.

File: src/axom/primal/tests/primal_intersect.cpp (hunk 2 above)
Comment (BradWhitlock) [reply src/axom/primal/tests/primal_intersect.cpp:291]: I had codex increase the test coverage for these intersections.

File: src/axom/primal/tests/primal_slice.cpp
Comment (BradWhitlock) [reply src/axom/primal/tests/primal_slice.cpp:1]: Added tests for these edge cases.

File: src/axom/primal/operators/slice.hpp (hunk 3 above)
Comment (BradWhitlock) [reply src/axom/primal/operators/slice.hpp:41]: The comments were improved.

File: src/axom/primal/operators/detail/intersect_impl.hpp (hunk 4 above)
Comment (BradWhitlock) [reply src/axom/primal/operators/detail/intersect_impl.hpp:142]: Done

File: src/axom/primal/operators/detail/slice_impl.hpp (hunk 5 above)
Comment (BradWhitlock) [reply src/axom/primal/operators/detail/slice_impl.hpp:132]: Yes. Done.

**Review by Arlie-Capps — APPROVED** [review]
Thank you, Brad!

**Review by bmhan12 — APPROVED** [review]

**Review by kennyweiss — APPROVED** [review]
Thanks for the improvements @BradWhitlock !
